lib/charms/smtp_integrator/v0/smtp.py

```diff
@@ -0,0 +1,286 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# Licensed under the Apache2.0. See LICENSE file in charm source for details.
+
+"""Library to manage the integration with the SMTP Integrator charm.
+
+This library contains the Requires and Provides classes for handling the integration
+between an application and a charm providing the `smtp` integration.
+It also contains a `SmtpRelationData` class to wrap the SMTP data that will
+be shared via the integration.
+
+### Requirer Charm
+
+```python
+
+from charms.smtp_integrator.v0 import SmtpDataAvailableEvent, SmtpRequires
+
+class SmtpRequirerCharm(ops.CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.smtp = smtp.SmtpRequires(self)
+        self.framework.observe(self.smtp.on.smtp_data_available, self._handler)
+        ...
+
+    def _handler(self, events: SmtpDataAvailableEvent) -> None:
+        ...
+
+```
+
+As shown above, the library provides a custom event to handle the scenario in
+which new SMTP data has been added or updated.
+
+### Provider Charm
+
+Following the previous example, this is an example of the provider charm.
+
+```python
+from charms.smtp_integrator.v0 import SmtpDataAvailableEvent, SmtpProvides
+
+class SmtpProviderCharm(ops.CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.smtp = SmtpProvides(self)
+        ...
+
+```
+The SmtpProvides object wraps the list of relations into a `relations` property
+and provides an `update_relation_data` method to update the relation data by passing
+a `SmtpRelationData` data object.
+"""
+
+# The unique Charmhub library identifier, never change it
+LIBID = "018a12e798714b69829be1e9a5c481a5"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+# pylint: disable=wrong-import-position
+import logging
+from enum import Enum
+from typing import Optional
+
+import ops
+from pydantic import BaseModel, Field
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "smtp-legacy"
```

Comment:
Out of curiosity, what is the reason for calling this legacy please?

Reply:
You can find the details in the spec

Reply:
The spec says it will provide two interfaces, one legacy and one "standard" or how you call it - I've noticed in the spec the interface is with an underscore - "smtp_legacy", it should be with an underscore here as well to match the spec, right? And as the charm will provide two interfaces, perhaps document that there is a legacy interface and how to use it, but go with the newer one as default? Still for the initial minimal working charm I think this can stay as is, however you prefer.